lib/level_db.rb

```diff
@@ -113,6 +113,24 @@ def each(options={}, &block)
     def snapshot
       Snapshot.new(@db)
     end
+
+    def compact_range(options={})
+      from = options[:from]
+      to = options[:to]
+      @db.compact_range(encode_key(from), encode_key(to))
+    end
+
+    def full_compaction
+      sstables = @db.get_property(SSTABLES_PROPERTY)
+      pre_compaction_sstables = nil
+      until pre_compaction_sstables == sstables
+        pre_compaction_sstables = sstables
+        compact_range
+        sstables = @db.get_property(SSTABLES_PROPERTY)
+      end
```

Comment:
Sounds like there are quite a few problems with this loop. It would be reasonable to assume that the order of the files in the `leveldb.sstables` property would be the same from call to call, but _reasonable_ is probably not enough. I wouldn't be comfortably with anything less than sorting the lists before comparing (I didn't even think of this on my first look at the code, so good that you pointed that out).

As for the level counting I don't know if that's better or worse. What if the same number of SSTables were added that were compacted away between two calls? I say that unless there's anything else wrong with looking at the files then there should be no problem looking at the list of files (provided it's sorted).

A max iterations parameter sounds like a must too. I would raise an error if there was no convergence after the max number of iterations.